[css-mixins-1] Parse @function body

anttijk opened the pull request:
#### bd44685c5370772e97308f946e7455831e726fd5
<pre>
[css-mixins-1] Parse @function body
<a href="https://bugs.webkit.org/show_bug.cgi?id=298095">https://bugs.webkit.org/show_bug.cgi?id=298095</a>
<a href="https://rdar.apple.com/159440523">rdar://159440523</a>

Reviewed by Sam Weinig and Matthieu Dubet.

<a href="https://drafts.csswg.org/css-mixins/#function-body">https://drafts.csswg.org/css-mixins/#function-body</a>

@function body accepts nested @media and other conditional rules along with custom properties so
the parsing is more like regular style rules.

Also add most of the CSSOM to get some test coverage, <a href="https://drafts.csswg.org/css-mixins/#cssom">https://drafts.csswg.org/css-mixins/#cssom</a>

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-mixins/at-function-cssom-expected.txt:

Type parameter serialization is not implemented yet.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/JSCSSRuleCustom.cpp:
(WebCore::toJSNewlyCreated):
* Source/WebCore/bindings/js/JSCSSStyleDeclarationCustom.cpp:
(WebCore::toJSNewlyCreated):
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/css/CSSFunctionDeclarations.cpp: Copied from Source/WebCore/css/StyleRuleFunction.cpp.
(WebCore::CSSFunctionDeclarations::CSSFunctionDeclarations):
(WebCore::CSSFunctionDeclarations::style):
(WebCore::CSSFunctionDeclarations::cssText const):
(WebCore::CSSFunctionDeclarations::reattach):
* Source/WebCore/css/CSSFunctionDeclarations.h: Copied from Source/WebCore/css/StyleRuleFunction.cpp.
* Source/WebCore/css/CSSFunctionDeclarations.idl: Copied from Source/WebCore/css/StyleRuleFunction.cpp.
* Source/WebCore/css/CSSFunctionDescriptors.cpp: Copied from Source/WebCore/css/StyleRuleFunction.cpp.
(WebCore::CSSFunctionDescriptors::CSSFunctionDescriptors):
(WebCore::CSSFunctionDescriptors::ruleType const):
(WebCore::CSSFunctionDescriptors::result const):
(WebCore::CSSFunctionDescriptors::setResult):
* Source/WebCore/css/CSSFunctionDescriptors.h: Copied from Source/WebCore/css/StyleRuleFunction.cpp.
* Source/WebCore/css/CSSFunctionDescriptors.idl: Copied from Source/WebCore/css/StyleRuleFunction.cpp.
* Source/WebCore/css/CSSFunctionRule.cpp: Added.
(WebCore::CSSFunctionRule::create):
(WebCore::CSSFunctionRule::CSSFunctionRule):
(WebCore::CSSFunctionRule::name const):
(WebCore::CSSFunctionRule::getParameters const):
(WebCore::CSSFunctionRule::returnType const):
(WebCore::CSSFunctionRule::cssText const):
(WebCore::CSSFunctionRule::styleRuleFunction const):
* Source/WebCore/css/CSSFunctionRule.h: Copied from Source/WebCore/css/StyleRuleFunction.cpp.
* Source/WebCore/css/CSSFunctionRule.idl: Copied from Source/WebCore/css/StyleRuleFunction.cpp.
* Source/WebCore/css/CSSStyleDeclaration.h:
* Source/WebCore/css/PropertySetCSSDescriptors.cpp:
(WebCore::PropertySetCSSDescriptors::getPropertyValue):
* Source/WebCore/css/StyleRule.cpp:
(WebCore::StyleRuleBase::createCSSOMWrapper const):
* Source/WebCore/css/StyleRule.h:
(WebCore::StyleRuleBase::isGroupRule const):
* Source/WebCore/css/StyleRuleFunction.cpp:
(WebCore::StyleRuleFunctionDeclarations::mutableProperties):
* Source/WebCore/css/StyleRuleFunction.h:
* Source/WebCore/css/parser/CSSParser.cpp:
(WebCore::CSSParser::consumeAtRule):

Add an allow mode to only allow &quot;conditional group rules&quot;.

(WebCore::CSSParser::consumeNestedGroupRules):
(WebCore::CSSParser::consumeFunctionRule):
(WebCore::CSSParser::consumeBlockContent):
(WebCore::CSSParser::consumeDeclarationRuleListInNewNestingContext):
(WebCore::CSSParser::consumeDeclarationList):
(WebCore::CSSParser::consumeDeclarationRuleList):

@function body takes &lt;declaration-rule-list&gt;, add parsing support for it.

(WebCore::CSSParser::consumeStyleBlock):
(WebCore::ruleDoesNotAllowImportant):
* Source/WebCore/css/parser/CSSParser.h:
(WebCore::CSSParser::isStyleNestedContext const):
(WebCore::CSSParser::isFunctionNestedContext const):
* Source/WebCore/css/parser/CSSParserEnum.h:
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::CSSPropertyParser::parseValue):
(WebCore::consumeFunctionDescriptor):
* Source/WebCore/css/query/MediaQueryParser.h:

Canonical link: <a href="https://commits.webkit.org/299544@main">https://commits.webkit.org/299544@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/825bab3953c62034cbbb25c6f4bf778f48e8f269

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/119222 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/38905 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/29559 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/125458 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/71288 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/9f8c51d7-429f-4321-954f-8dd601fa7fb5) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/121100 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/39602 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/47487 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/90580 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/59940 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/f9203166-95b0-4c15-9214-77dd61110d51) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/122175 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/31580 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/106881 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/70989 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/3709a776-d62c-4c99-9ac5-d0375660c523) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/30624 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/24994 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/69103 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/101027 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/25179 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/128466 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/46134 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/34874 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/99140 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/46499 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/103091 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/98915 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25175 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/44385 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/22392 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/42711 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/46001 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/51684 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/45468 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/48818 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/47158 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

mdubet commented on Source/WebCore/css/parser/CSSParser.cpp:
```diff
@@ -1495,7 +1497,9 @@ RefPtr<StyleRuleBase> CSSParser::consumeStyleRule(CSSParserTokenRange prelude, C
-        return isStyleNestedContext() && onlyDeclarations == OnlyDeclarations::No;
+        if (onlyDeclarations == OnlyDeclarations::Yes)
+            return false;
+        return isStyleNestedContext() || isFunctionNestedContext();
```
`@function` body is a [declaration-rule-list](https://drafts.csswg.org/css-syntax-3/#typedef-declaration-rule-list)

It seems the current patch allow qualified rule (~style rule) in the body ?

mdubet reviewed (changes requested):
From reading the patch, it seems style rule would be parsed (and stored) inside `@function` body which is wrong.

We should add an enum parameter to consumeBlockContent with the spec wording (at least BlockContent/QualifiedRuleList for the current patch) to decide what is accepted (descriptors and/or at-rules and/or qualified-rules).

```cpp
enum class BlockType {
	BlockContent
	QualifiedRuleList
	....etc
}
```

Also it's surprising that this patch doesn't affect any test ?

mdubet commented on Source/WebCore/css/parser/CSSParser.cpp:
```diff
@@ -1492,12 +1511,8 @@ RefPtr<StyleRuleBase> CSSParser::consumeStyleRule(CSSParserTokenRange prelude, C
 // https://drafts.csswg.org/css-syntax/#consume-block-contents
 // https://drafts.csswg.org/css-syntax/#block-contents
-void CSSParser::consumeBlockContent(CSSParserTokenRange range, StyleRuleType ruleType, OnlyDeclarations onlyDeclarations, ParsingStyleDeclarationsInRuleList isParsingStyleDeclarationsInRuleList)
+void CSSParser::consumeBlockContent(CSSParserTokenRange range, StyleRuleType ruleType, OptionSet<BlockAllowedRule> blockAllowedRules, ParsingStyleDeclarationsInRuleList isParsingStyleDeclarationsInRuleList)
```
It seems we are never checking the Declarations option ?

anttijk replied:
Right because all branches currently in use support Declarations (as you note below).

mdubet commented on Source/WebCore/css/parser/CSSParser.cpp:
```diff
@@ -1635,14 +1659,33 @@ ParsedPropertyVector CSSParser::consumeDeclarationListInNewNestingContext(CSSPar
 {
-    consumeBlockContent(range, ruleType, OnlyDeclarations::No, isParsingStyleDeclarationsInRuleList);
+    // <rule-list>: qualified rules and at-rules are allowed; declarations are automatically invalid.
+    consumeBlockContent(range, ruleType, { BlockAllowedRule::QualifiedRules, BlockAllowedRule::AtRules }, isParsingStyleDeclarationsInRuleList);
```
StyleBlock supports everything : declarations + qualified rules + at-rules

anttijk replied:
Yeah copied the wrong comment.

anttijk committed:
[css-mixins-1] Parse @function body
https://bugs.webkit.org/show_bug.cgi?id=298095
rdar://159440523

Reviewed by Sam Weinig and Matthieu Dubet.

https://drafts.csswg.org/css-mixins/#function-body

@function body accepts nested @media and other conditional rules along with custom properties so
the parsing is more like regular style rules.

Also add most of the CSSOM to get some test coverage, https://drafts.csswg.org/css-mixins/#cssom

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-mixins/at-function-cssom-expected.txt:

Type parameter serialization is not implemented yet.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/JSCSSRuleCustom.cpp:
(WebCore::toJSNewlyCreated):
* Source/WebCore/bindings/js/JSCSSStyleDeclarationCustom.cpp:
(WebCore::toJSNewlyCreated):
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/css/CSSFunctionDeclarations.cpp: Copied from Source/WebCore/css/StyleRuleFunction.cpp.
(WebCore::CSSFunctionDeclarations::CSSFunctionDeclarations):
(WebCore::CSSFunctionDeclarations::style):
(WebCore::CSSFunctionDeclarations::cssText const):
(WebCore::CSSFunctionDeclarations::reattach):
* Source/WebCore/css/CSSFunctionDeclarations.h: Copied from Source/WebCore/css/StyleRuleFunction.cpp.
* Source/WebCore/css/CSSFunctionDeclarations.idl: Copied from Source/WebCore/css/StyleRuleFunction.cpp.
* Source/WebCore/css/CSSFunctionDescriptors.cpp: Copied from Source/WebCore/css/StyleRuleFunction.cpp.
(WebCore::CSSFunctionDescriptors::CSSFunctionDescriptors):
(WebCore::CSSFunctionDescriptors::ruleType const):
(WebCore::CSSFunctionDescriptors::result const):
(WebCore::CSSFunctionDescriptors::setResult):
* Source/WebCore/css/CSSFunctionDescriptors.h: Copied from Source/WebCore/css/StyleRuleFunction.cpp.
* Source/WebCore/css/CSSFunctionDescriptors.idl: Copied from Source/WebCore/css/StyleRuleFunction.cpp.
* Source/WebCore/css/CSSFunctionRule.cpp: Added.
(WebCore::CSSFunctionRule::create):
(WebCore::CSSFunctionRule::CSSFunctionRule):
(WebCore::CSSFunctionRule::name const):
(WebCore::CSSFunctionRule::getParameters const):
(WebCore::CSSFunctionRule::returnType const):
(WebCore::CSSFunctionRule::cssText const):
(WebCore::CSSFunctionRule::styleRuleFunction const):
* Source/WebCore/css/CSSFunctionRule.h: Copied from Source/WebCore/css/StyleRuleFunction.cpp.
* Source/WebCore/css/CSSFunctionRule.idl: Copied from Source/WebCore/css/StyleRuleFunction.cpp.
* Source/WebCore/css/CSSStyleDeclaration.h:
* Source/WebCore/css/PropertySetCSSDescriptors.cpp:
(WebCore::PropertySetCSSDescriptors::getPropertyValue):
* Source/WebCore/css/StyleRule.cpp:
(WebCore::StyleRuleBase::createCSSOMWrapper const):
* Source/WebCore/css/StyleRule.h:
(WebCore::StyleRuleBase::isGroupRule const):
* Source/WebCore/css/StyleRuleFunction.cpp:
(WebCore::StyleRuleFunctionDeclarations::mutableProperties):
* Source/WebCore/css/StyleRuleFunction.h:
* Source/WebCore/css/parser/CSSParser.cpp:
(WebCore::CSSParser::consumeAtRule):

Add an allow mode to only allow "conditional group rules".

(WebCore::CSSParser::consumeNestedGroupRules):
(WebCore::CSSParser::consumeFunctionRule):
(WebCore::CSSParser::consumeBlockContent):
(WebCore::CSSParser::consumeDeclarationRuleListInNewNestingContext):
(WebCore::CSSParser::consumeDeclarationList):
(WebCore::CSSParser::consumeDeclarationRuleList):

@function body takes <declaration-rule-list>, add parsing support for it.

(WebCore::CSSParser::consumeStyleBlock):
(WebCore::ruleDoesNotAllowImportant):
* Source/WebCore/css/parser/CSSParser.h:
(WebCore::CSSParser::isStyleNestedContext const):
(WebCore::CSSParser::isFunctionNestedContext const):
* Source/WebCore/css/parser/CSSParserEnum.h:
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::CSSPropertyParser::parseValue):
(WebCore::consumeFunctionDescriptor):
* Source/WebCore/css/query/MediaQueryParser.h:

Canonical link: https://commits.webkit.org/299544@main
